### ISSUE TITLE
TT-Tomography delivering wrong results and raypaths due to topo ?

### ISSUE THREAD

**Issue by GGDRriedel** [gimli-org/gimli#402]
### Problem description

Please describe your issue here.
I am doing traveltime tomography on a hilly profile. However, raytracing (and the velocity model) are delivering impossible results. 

### Your environment

**Operating system**: tested on ubuntu 20.04 and Windows 10
**Python version**:  Win: 3.8 , Ubuntu: 3.7
**pyGIMLi version**: LINUX: 1.2.1, Windows 1.11
**Way of installation**: Conda  & Conda

### Steps to reproduce

Tell us how to reproduce this issue. Ideally, you could paste the code that produces the error:

data:
[p1.txt](https://github.com/gimli-org/gimli/files/8998841/p1.txt)

```python
import matplotlib.pyplot as plt
import numpy as np
import pygimli as pg
from pygimli.physics import Refraction

profile="p1.sgt"
ras = Refraction(profile)
mesh=ras.createMesh(data=ras.data,paraDepth=150,quality=32,area=20)
#fig=ras.showMesh()
ras.invert(data=ras.data,useGradient=True,vTop=1000,vBottom=4000,secNodes=3, mesh=mesh,verbose=True)  
ax, cbar = ras.showResult(logScale=False, cMap="viridis",
                      coverage=ras.standardizedCoverage())
ras.drawRayPaths(ax=ax, color="k", lw=0.3, alpha=0.5)
```

### Expected behavior
result from different software: 
![grafik](https://user-images.githubusercontent.com/48824691/176124089-abfd3c21-d66f-48ce-a103-798820f32776.png)

### Actual behavior
raypaths are not taking an expected path in topographical areas
![grafik](https://user-images.githubusercontent.com/48824691/176181390-63ea5496-eac7-4419-a50c-a8ca67acf85f.png)
```


```

So obviously the raypaths on that hill are all wrong, going along the very very top cells instead of through the subsurface. 

Help would be apreciated. maybe it's just some parameter I am missing?

**Comment by halbmy** [comment]
Well the ray paths are not wrong but a result of the model that shows a velocity decrease with depth. Consequently there is no sensitivity and therefore the model stays close to the starting (gradient) model. Maybe the starting model generation goes wrong. 

I've noticed that both pyGIMLi versions are quite old. Did you check with a recent version?

**Comment by halbmy** [comment]
Well it seems like our gradient model generation routine is not quite suited for such topographic undulations:
```
data = tt.load("p1.sgt")
ras = tt.TravelTimeManager(data)
mesh=ras.createMesh(data=data,paraDepth=150,quality=34,area=20)
slo = tt.createGradientModel2D(ras.data, mesh, vTop=1000, vBot=4000)
pg.show(mesh, 1./slo, cMap="plasma_r")
```
![image](https://user-images.githubusercontent.com/7136401/176276899-0e7dc1ab-1518-46df-b1dd-e5203f49c365.png)

However if you create your own starting model
```
cz = pg.y(mesh.cellCenters())
vel = 1000 + (max(cz)-cz)/(max(cz)-min(cz)) * 3000
pg.show(mesh, vel, cMap="plasma_r")
```
it should be better suited.

At any rate, I observed zero or near-zero offsets in your data that should better be removed to avoid extremely fine meshes:
...
15.57	421.69
15.58	421.69
...
5	6	7.000000000000001e-05
...
```

**Comment by halbmy** [comment]
Actually, the better solution is the distance from the surface:
```
cx = pg.x(mesh.cellCenters())
cz = pg.y(mesh.cellCenters())
cd = np.interp(cx, pg.x(data), pg.y(data)) - cz
vel = 1000 + cd / max(cd) * 3000
pg.show(mesh, vel, cMap="plasma_r")
...
ras.invert(..., startModel=1/vel)
```
![image](https://user-images.githubusercontent.com/7136401/176282747-02f892e6-fba9-4f21-82a8-2bc0d2ba67d1.png)

**Comment by halbmy** [comment]
Adapted `createGradientModel2D` to follow the topography (or keep the old behaviour with flat=True) #3d2b3e5

**Comment by GGDRriedel** [comment]
Thank you for the comprehensive answers!

The almost zero offset positions are a result of hammer blows right next to sensors but not quite and the field crew being very thorough with their work :)

I'll try to use the hand generated starting model as I am not able to update to the newest version as of right now. 

Also I plan on using this awesome GUI for Tomography bysed on PyGimli: 
https://github.com/viictorjs/Refrapy
Just to tell you about it ;)


I'll get back to you once it's finished the inversion

**Comment by halbmy** [comment]
Anyway you should get rid of your quasi-duplicate sensors as this will slow down the inversion heavily. Once you did this you can use a better mesh quality and refinement at the surface.

**Comment by GGDRriedel** [comment]
Much better results already:
![grafik](https://user-images.githubusercontent.com/48824691/176425983-fd39eba8-2e13-4055-a311-2a1c5f19ccd0.png)

**Comment by halbmy** [comment]
So how does the final result look like? It is comparable to the other software? If yes, we can close the issue.

**Comment by GGDRriedel** [comment]
Absolutely. Better results, better output options and with Pyvista + Paraview I can show crossing Tomolines in 3D , translated to the path  and have a myriad of export options. 

Great stuff. 

![grafik](https://user-images.githubusercontent.com/48824691/177101811-8380e678-c7f7-4efe-9f00-5ef177db5044.png)

**Comment by halbmy** [comment]
Great to hear! As the starting model function had been updated already we can close the issue. Thank's for reporting.
